magma/circuit.py

```diff
@@ -921,3 +923,32 @@ def __new__(metacls, name, bases, dct):
 
 class DebugCircuit(Circuit, metaclass=DebugDefineCircuitKind):
     pass
+
+
+InstanceCallback = Callable[[List[CircuitType], Dict], None]
+
+
+def register_instance_callback(
+        instance: CircuitType,
+        callback: InstanceCallback,
+):
+    try:
+        get_instance_callback(instance)
+    except AttributeError:
+        pass
+    else:
+        raise AttributeError(
+            f"Instance {instance} already has callback registered"
+        )
+    setattr(instance, "_callback_", callback)
+
+
+def get_instance_callback(instance: CircuitType) -> InstanceCallback:
+    try:
+        return getattr(instance, "_callback_")
```

Comment:
```suggestion
        return instance._callback_"
```

Could skip getattr since we're catching the AttributeError anyways

Reply:
:+1: